### PR TITLE
Port //ledger/ledger-api-client/... to Scala 2.13

### PR DESCRIPTION
**Pull request by cocreature**
This pulls in Sandbox next and kvutils as a dependency so those now
build on 2.13 as well.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala
Comment (cocreature) [inline ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala:156]: symbol literals have been deprecated.

File: ledger/ledger-api-client/src/test/2.12/com/daml/ledger/client/services/commands/Compat.scala
Comment (cocreature) [inline ledger/ledger-api-client/src/test/2.12/com/daml/ledger/client/services/commands/Compat.scala:4]: this just got shuffled around since we don’t need it on 2.13.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/queries/Queries.scala
Comment (cocreature) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/queries/Queries.scala:40]: you need an immutable Seq now for varargs.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Envelope.scala
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Envelope.scala:96]: either is right biased and warns about usage of `.right` in 2.13.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:71]: I couldn’t find a nice way of making this work in the previous generality on both 2.12 and 2.13 and we only use it for `Map[K, V]` so I’ve just restricted the type.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:160]: `Traversable` is dead in favor of `Iterable`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/LedgerDataImporter.scala
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/LedgerDataImporter.scala:9]: `Stream` is deprecated.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/package.scala
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/package.scala:12]: `Seq` defaults to `immutable.Seq` now but we put buffers in here.

File: bazel-java-deps.bzl
Comment (cocreature) [inline bazel-java-deps.bzl:141]: The old version suffers from an issue in `LazyList` that results in a deadlock in `//ledger/participant-state/kvutils:kvutils-tests_test_suite_src_test_suite_scala_com_daml_
ledger_participant_state_kvutils_export_ProtobufBasedLedgerDataExportSpec.scala`. Afaict, the commit fixing this is https://github.com/scala/scala-collection-compat/pull/350/files but I haven’t looked too closely at the actual fixes upstream.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Debug.scala
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Debug.scala:9]: Why hiding `Envelope`?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/StateSerializationStrategy.scala
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/StateSerializationStrategy.scala:22]: I assume here it was just impossible to do something like `.to(SortedMap)`, right?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala:196]: 🧐

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala:106]: Doesn't make sense to fix in this PR, but sorting a ~stream~ lazy list smells a bit.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala:56]: This _may_ be an improvement. We tried to remove usages of `.mapValues` in the past because we found out it was recomputing it at every usage of the resulting collection. I believe the new Collections API should have fixed it.

**Review by stefanobaghino-da — APPROVED** [review]
LGTM but may be worth to wait for a second review from an owner of the `kvutils` part.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Debug.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Debug.scala:9]: We want `com.daml.ledger.participant.kvutils.Envelope` below. The scoping rules in 2.13 seem to be different which means that we end up with the one from `DamlKvutils` otherwise.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/StateSerializationStrategy.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/StateSerializationStrategy.scala:22]: I’ve tried that first and it works on Scala 2.13 but I couldn’t get it to typecheck on 2.12 so at some point I gave up and went down this route.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala:106]: good point, I’ll leave that to someone who actually understands kvutils :slightly_smiling_face:

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala:56]: yeah the new API makes more sense to me.

**Comment by cocreature** [comment]
> LGTM but may be worth to wait for a second review from an owner of the `kvutils` part.

Yep I’ll hold off on merging until I have a review from a codeowner.

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala
Comment (cocreature) [reply ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala]: No good reason, I just went for the mechanical transformation (regex replace \o/) instead of thinking about it as I should have. fixed in a33ebf455c78bb350594433fc466cde27b8056fd

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/queries/Queries.scala
Comment (cocreature) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/queries/Queries.scala]: fixed in 5bb65640f53700a55ca35e429558bafe08477dfb

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala:103]: fixed in af35d721baa0b5e440089c0b1bcd5c5d3c5915f2

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala:56]: https://github.com/scala/scala/blob/v2.13.4/src/library/scala/collection/IterableOnce.scala#L1255 calls `Map.from` which forces the thing. iirc the deprecation warning mentions that they will add a strict version of `mapValues` in the future. Haven’t checked if it’s in Scala 3.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/package.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/package.scala:12]: `mutable.Seq` means you can actually mutate it whereas `collection.Seq` just means that the underlying data structure can be mutable but you’re not going to mutate it. I find mutability scary so I’d rather keep `mutable.Seq` for cases where we actually mutate things.

**Pull request merged**
